# Step 1 - Khảo sát dữ liệu Wine Quality Red

Phần này đọc dữ liệu và kiểm tra các thông tin cơ bản theo yêu cầu của bài.

In [ ]:
import pandas as pd

file_path = "../data/D09 - winequality-red.xlsx"

df = pd.read_excel(
    file_path,
    sheet_name="winequality-red"
)

df.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


### 1a. Kích thước và chiều của dữ liệu

In [2]:
print("Kích thước:", df.shape)
print("Số chiều:", df.ndim)
print("Số dòng:", df.shape[0])
print("Số cột:", df.shape[1])

Kích thước: (1599, 12)
Số chiều: 2
Số dòng: 1599
Số cột: 12


Dữ liệu có **1599 mẫu và 12 cột**. Trong đó có 11 cột thuộc tính và cột `quality` là nhãn.

### 1b. Kiểu dữ liệu của các thuộc tính

In [3]:
df.dtypes

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

Các thuộc tính đầu vào đều là dữ liệu số thực (`float64`), còn `quality` là số nguyên (`int64`).

### 1c. Số lượng mẫu của từng giá trị nhãn

In [4]:
label_counts = df["quality"].value_counts().sort_index()
label_counts

quality
3     10
4     53
5    681
6    638
7    199
8     18
Name: count, dtype: int64

Có 6 mức chất lượng từ 3 đến 8. Nhãn 5 và 6 có nhiều mẫu nhất, còn nhãn 3 và 8 có khá ít mẫu nên phân bố nhãn chưa cân bằng.

### 1d. Giá trị nhỏ nhất, lớn nhất và trung bình của các thuộc tính

In [5]:
features = df.drop(columns="quality")

statistics = features.agg(["min", "max", "mean"]).T
statistics.round(3)

,min,max,mean
fixed acidity,4.600,15.900,8.320
volatile acidity,0.120,1.580,0.528
citric acid,0.000,1.000,0.271
residual sugar,0.900,15.500,2.539
chlorides,0.012,0.611,0.087
free sulfur dioxide,1.000,72.000,15.875
total sulfur dioxide,6.000,289.000,46.468
density,0.990,1.004,0.997
pH,2.740,4.010,3.311
sulphates,0.330,2.000,0.658


Các thuộc tính có thang giá trị khá khác nhau. Ví dụ `density` có giá trị quanh 1, trong khi `total sulfur dioxide` có thể lớn hơn 200.

### Kiểm tra thêm: giá trị thiếu

In [6]:
df.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

Không có giá trị thiếu trong bộ dữ liệu.